kw6/reader.py

```diff
@@ -20,61 +20,118 @@ class Reader(BaseModel):
-        stream = Path(path).open('rb')
+    @validate_arguments
+    def from_path(path: Path, header_path: Path = None):
+        stream = io.BufferedReader(path.open("rb"))
+
         version = stream.read(settings.N_BYTES_VERSION).decode().strip()
-        if version != 'KW6FileClassVer1.0':
-            raise ValueError(f'Unexpected file version {version}')
-        n_bytes_position, initial_frame_index = initial_position_info(stream)
+        if version != "KW6FileClassVer1.0":
+            raise ValueError(f"Unexpected file version {version}")
+
+        initial_position_header = Position.skip_(stream)
         return Reader(
+            path=path,
             stream=stream,
-            cached_positions=header_positions(header_path) if header_path is not None else {},
-            n_bytes_position=n_bytes_position,
-            initial_frame_index=initial_frame_index,
+            cached_positions=(
+                header_positions(header_path)
+                if header_path is not None
+                else dict()
+            ),
+            initial_frame_index=initial_position_header.frame_index,
         )
 
+    def __iter__(self):
+        """Iterate over all positions and cameras in the file"""
+        stream = io.BufferedReader(self.path.open("rb"))
+        stream.seek(settings.N_BYTES_VERSION)
+        while not stream.peek(1) == b"":
+            byte_position = stream.tell()
+            position = Position.from_stream_(stream)
+            self.cached_positions[position.header.frame_index] = byte_position
+            yield position
+        stream.close()
+
+    def __len__(self):
+        try:
+            return self.assumptuous_length()
+        except Exception:
+            self.stream.seek(settings.N_BYTES_VERSION)
+            if self.empty():
+                return 0
+
+            position_header = None
+            while not self.empty():
+                byte_position = self.stream.tell()
+                position_header = Position.skip_(self.stream)
+                self.cached_positions[position_header.frame_index] = byte_position
+
+            return position_header.frame_index - self.initial_frame_index + 1
+
+    def assumptuous_length(self):
+        self.stream.seek(settings.N_BYTES_VERSION)
+        position_header = self.position_header()
+        self.stream.seek(0, io.SEEK_END)
+        end_position = self.stream.tell()
+        assumptuous_length = (
+            end_position - settings.N_BYTES_VERSION
+        ) / position_header.n_frame_bytes
+        print("assumptuous_length:", assumptuous_length)
+
+        if np.abs(assumptuous_length - np.round(assumptuous_length)) <= 1e-9:
+            assumptuous_length = int(assumptuous_length)
+        else:
+            raise Exception(f"Unexpected length {assumptuous_length} of kw6 file")
+
+        expected_frame_index = self.initial_frame_index + assumptuous_length - 1
+        position = self[expected_frame_index]
```

Comment:
Isn't this very expensive if we are unable to do an assumptuous seek and don't have a header file?

Reply:
Yes, this is a miss. I was thinking I was just doing a seek + read manually but this is calling a lot more things